tools/model-apt/src/main/java/io/smallrye/openapi/model/apt/SmallRyeModelProcessor.java

```diff
@@ -0,0 +1,743 @@
+            buffer.setLength(0);
+            buffer.append(property.singularName);
+            buffer.setCharAt(0, Character.toUpperCase(property.singularName.charAt(0)));
+            property.singularMethodName = buffer.toString();
+        }
+
+        return property;
+    }
+
+    private String getSingularName(String propertyName, Map<String, Object> values) {
+        String singularName = get(values, "singularName", String.class);
+
+        if (singularName.isBlank()) {
+            if (propertyName.endsWith("s")) {
+                singularName = propertyName.substring(0, propertyName.length() - 1);
+            } else {
+                throw new IllegalArgumentException(
+                        "singularName is required for map or list type when name does not end with 's': " + propertyName);
+            }
+        }
+
+        return singularName;
+    }
+
+    private void writeProperty(Writer writer, String className, PropertyInfo property)
+            throws IOException {
+
+        if (isCentralizedProperty(property)) {
+            return;
+        }
+
+        writeGet(writer, property);
+        writeSet(writer, property);
+
+        if (Map.class.getName().equals(property.rawType)) {
+            writeMapAdd(writer, className, property);
+            writeMapRemove(writer, property);
+        } else if (List.class.getName().equals(property.rawType)) {
+            writeListAdd(writer, className, property);
+            writeListRemove(writer, property);
+        }
+    }
+
+    /**
+     *
+     * @param writer
+     * @param property
+     * @throws IOException
+     */
+    private void writeGet(Writer writer, PropertyInfo property) throws IOException {
+        writeJavaDoc(writer);
+        writeOverride(writer);
+
+        writeCodeLn(writer, 1, PUBLIC, property.type, " get", property.methodName, "() {");
+
+        if (property.unwrapped) {
+            String valueTypeParam;
+
+            if (property.valueType.indexOf('<') > -1) {
+                /*
+                 * We only care about the raw type, but the compiler needs to be fooled to obtain
+                 * the type with generic arguments to satisfy the call to getProperties, and the
+                 * return type.
+                 */
+                String rawType = property.valueType.substring(0, property.valueType.indexOf('<'));
+                writeCodeLn(writer, 2, "@SuppressWarnings(\"unchecked\")");
+                writeCodeLn(writer, 2, "Class<", property.valueType, "> TYPE = (Class<", property.valueType, ">) (Object) ",
+                        rawType,
+                        CLASS, ";");
+                valueTypeParam = "TYPE";
+            } else {
+                valueTypeParam = property.valueType + CLASS;
+            }
+
+            writeCodeLn(writer, 2, "return getProperties(", valueTypeParam, ");");
+        } else {
+            writeCodeLn(writer, 2, "return ", property.getPropertyMethod, "(\"", property.name, "\");");
```

Comment:
I think this generates a call to `BaseExtensibleModel.getProperty(String)` which does not do any type checking (vs `getProperty(String, Class<?>)` which returns `null` if the property is not the right type.

I guess that's ok as long as we're certain that what's in the map is valid (e.g. we validate it on read). What's the overall strategy for ensuring we don't get `ClassCastException`s at runtime if the user feeds us an invalid openapi.yaml?

Reply:
Generally, we should be protected from invalid input by the expected types declared via the `DataType` instances. The generated `SmallRyeOASModels` class contains a map of all `Constructible`s as keys, and values containing the known properties and their `DataType`s. It ends up being similar to the logic that was in `SchemaIO`.

Reply:
For `Schema` we did two things:
1. In `SchemaIO` when reading, try to read the JSON as the desired type
  - fall back to reading it as a basic type if it can't be read as a desired type 
2. In `SchemaImpl` when getting a value, check if it's the correct type before returning it
  - if it's not the correct type, return `null`
  - The actual value could still be obtained with `Schema.get(String)`

I think this part is doing the getting, and unlike `SchemaImpl`, it isn't checking the type before returning the value, resulting in a `ClassCastException` when Java checks the return type of the generated getter method against the value being returned if it's not of the expected type.

Reply:
I think we can either implement the same logic in the getters as `SchemaImpl` did, or ensure that any invalid data is dropped when reading as I think the old IO classes would do.

Reply:
I think +1 to dropping invalid data in IO as well as the type checking on the getters. I think we are close to that in the current IO code, with the exception of the class cast bug you noted above. I'll put up a PR with the changes.